### PR TITLE
Add some emoji spinners

### PR DESCRIPTION
**Pull request by nnmrts**
I played around with some emoji and emoji sequences, and this is the result.

### CODE REVIEW

File: spinners.json [hunk 1]
```diff
@@ -1181,5 +1181,165 @@
...
+		"frames": [
+			"☝️",
+			"✌️",
+			"🤘",
+			"🤟",
+			"🖖",
+			"🖐",
+			"✋",
+			"🤚",
+			"👆"
```
Comment (sindresorhus) [inline spinners.json:1196]: Seems like these are not all the same rendered width. I tried it in `example.js` and the `Unicorn` text moves.

File: spinners.json [hunk 2]
```diff
@@ -1181,5 +1181,165 @@
 			"βββββρβ",
 			"ββββββρ"
 		]
+	},
+	"fingerdance": {
```
Comment (sindresorhus) [inline spinners.json:1185]:
```suggestion
	"fingerDance": {
```

File: spinners.json [hunk 3]
```diff
@@ -1181,5 +1181,165 @@
...
+		]
+	},
+	"fistBump": {
+		"interval": 80,
+		"frames": [
+			"🤜    🤛",
+			"🤜    🤛",
+			"🤜    🤛",
+			" 🤜  🤛",
+			"  🤜🤛",
```
Comment (sindresorhus) [inline spinners.json:1206]: The width should be constant.

Comment (sindresorhus) [reply spinners.json:1206]: This one is awesome though!

File: spinners.json [hunk 4]
```diff
@@ -1181,5 +1181,165 @@
...
+	"header": {
+		"interval": 80,
+		"frames": [
+			" 🧑⚽️       🧑",
+			"🧑  ⚽️      🧑",
+			"🧑   ⚽️     🧑",
+			"🧑    ⚽️    🧑",
+			"🧑     ⚽️   🧑",
+			"🧑      ⚽️  🧑",
+			"🧑       ⚽️🧑",
```
Comment (sindresorhus) [inline spinners.json:1220]: Constant width.

Can you write a test for that? To ensure that all the frames in each spinner has the same width.

File: spinners.json [hunk 5]
```diff
@@ -1181,5 +1181,165 @@
...
+			"🤜    🤛",
+			"🤜    🤛",
+			"🤜    🤛",
+			" 🤜  🤛",
+			"  🤜🤛",
+			" 🤜✨🤛",
+			"🤜 ✨ 🤛"
+		]
+	},
+	"header": {
```
Comment (sindresorhus) [inline spinners.json:1211]: I think the name could be better and more specific.

**Comment by sindresorhus** [comment]
These are all very nice! Thanks for making them. 🙌

File: spinners.json (hunk 3 above)
Comment (nnmrts) [reply spinners.json:1206]: Hmmm, I'm really struggling with this one. Is there a unicode character for a space with the width of an emoji?

File: spinners.json (hunk 5 above)
Comment (nnmrts) [reply spinners.json:1211]: Alright, I'll change it to "soccerHeader".

File: spinners.json [hunk 6]
```diff
@@ -1181,5 +1181,165 @@
...
+			"😧",
+			"😧",
+			"🤯",
+			"💥",
+			"✨",
+			"",
+			"",
+			""
+		]
+	},
```
Comment (nnmrts) [inline spinners.json:1246]: I'm also not sure about that one tbh. Yeah, it's "just" emoji and "just" a cli-spinner, but it looks pretty graphic and insensible. I think the same about the already implemented "grenade" spinner btw.

File: spinners.json (hunk 1 above)
Comment (nnmrts) [reply spinners.json:1196]: I couldn't try it out in my shell yet, but at least [in the browser it looks fine](https://jsfiddle.net/sg2whe7t/), is this a shell-only thing, that different emoji have different width?

Comment (nnmrts) [reply spinners.json:1196]: Okay, so for some reason ☝️ and ✌️ are not shown in their "emoji form", in my shell at least. Removing them and adding a space after the 🖐 one fixed it for me.

Kapture 2020-06-28 at 8 02 44

File: spinners.json (hunk 4 above)
Comment (nnmrts) [reply spinners.json]: Adding a test for constant width could be pretty hard, if not impossible. Look at this small example of the fingerDance spinner:

Kapture 2020-06-28 at 8 57 29

No matter which package or method I use, the number is never the same across all frames, even though visually every frame has the same width.

I used this code to log this btw:

<details>
<summary>code</summary>

```js
const logUpdate = require('log-update');
const cliSpinners = require('.');
const stringWidth = require("string-width");
const stringLength = require("string-length");

const spinner = cliSpinners[process.argv[2] || 'dots'];
let i = 0;

setInterval(() => {
	const {frames} = spinner;
	logUpdate(
		'\n' +
		frames[i = ++i % frames.length] + ' Unicorns' + 
		'\n' +
		'js length:\t\t' + frames[i].length +
		'\n' +
		'"string-width" length:\t' + stringWidth(frames[i]) +
		'\n' + 
		'"string-length" length:\t' + stringLength(frames[i])
	);
}, spinner.interval);
```


</details>

File: spinners.json (hunk 6 above)
Comment (sindresorhus) [reply spinners.json:1246]: I like this one. It's fun.

File: spinners.json (hunk 4 above)
Comment (sindresorhus) [reply spinners.json]: > Adding a test for constant width could be pretty hard, if not impossible. Look at this small example of the fingerDance spinner:

If the problem is just with that spinner, we could exclude it from that test.

File: spinners.json (hunk 6 above)
Comment (nnmrts) [reply spinners.json:1246]: Okay then :)

**Comment by sindresorhus** [comment]
If it was not clear, this is waiting for https://github.com/sindresorhus/cli-spinners/pull/46#discussion_r446803883

**Comment by sindresorhus** [comment]
@nnmrts Bump :)

**Comment by nnmrts** [comment]
@sindresorhus, ye, sorry, I'll look into it again tonight.

**Comment by sindresorhus** [comment]
@nnmrts Bump :)

**Comment by nnmrts** [comment]
tl;dr everything should be fixed now, some things were removed because they are dumb, the grenade spinner was updated as well because its first two frames were shorter and i added spaces at the end of all the emoji frames because the existing emoji spinners also have one space at the end; these spinners all have constant widths **in a terminal**, but not always in browsers; if that's an issue i'll remove those ones too

---

So, revisiting this, I don't really know how to continue in a sound way. Excuse my language but it seems like string and character width handling in the context of emoji is a complete clusterfuck. Different browsers, different terminals, different programming languages and different libraries all do it differently. I can't even visually compare these spinners in my IDE because VSCode **also** thinks for some reason it's a good idea that emoji have a slightly larger width than any other characters in an environment that specifically uses fixed-width characters, Terminus and the macOS Terminal thinks that too, Hyper on the other hand doesn't. All this probably changes with different fonts as well. This is literally untestable using normal spaces and emoji.

Anyways, considering this, and that this projects targets are mainly terminals, I added a constant width test using `string-length` that seems to be solid at checking the visual constant width of strings for all terminals. I also replaced "normal" spaces with [ideographic spaces](https://en.wikipedia.org/wiki/Whitespace_character#Unicode) where it's necessary. I copy-pasted them literally, because then you can actually visually see and compare the real lengths of these strings. This may lead to confusion though, I can also change them to `\u+3000` to make it clearer these aren't normal spaces, if you want me to.

Of course this wasn't enough, for some reason ☝️, ✌️ and 🖐️ don't have the usual emoji width in the terminals I tested, they are about a normal space width shorter. They have the usual width in VSCode and browsers though. They also aren't [ZWJ sequences](https://emojipedia.org/emoji-zwj-sequence/) or [keycap sequences](https://emojipedia.org/emoji-keycap-sequence/). I now removed those emoji from that spinner. This makes no sense at all though. At least with emoji like 1️⃣, it makes some sense that they behave different, because they are keycap sequences, but it still doesn't make sense that their emoji representations have completely different widths.

Oh shoot, did I mention that the math spinner is also completely broken and unfixable? I didn't? Well, the math spinner is also completely broken and unfixable. One could theoretically implement a parity test for keycap sequences (if there's an even number of these fuckup emoji in a string, they cancel each other out and the resulting string width should be predictable again), but I just called it a day at this point and completely removed this beautiful math spinner. Sad.

I never want to do this again, I regret this whole PR and I wish all the implementers of all of these text editors and displayers all the luck in the world figuring this out.

Thanks for coming to my ted talk.

File: spinners.json (hunk 1 above)
Comment (nnmrts) [reply spinners.json]: I've ultimately removed these emoji, so this seems to be resolved.

File: spinners.json (hunk 4 above)
Comment (nnmrts) [reply spinners.json]: Excluding the spinner from the test would've been somewhat of a solution, although the "fix" I had above doesn't work for every terminal I tested. I've removed the specific emoji causing the issue and added a test, so I think this is resolved as well.

File: test.js [hunk 7]
```diff
@@ -1,8 +1,29 @@
...
 
 test('main', t => {
 	t.is(typeof cliSpinners, 'object');
 	t.is(cliSpinners.dots.interval, 80);
 	t.true(Array.isArray(cliSpinners.dots.frames));
 });
+
+test('constant width', t => {
+	for (const key of Object.keys(cliSpinners)) {
+		if (key !== 'default') {
```
Comment (sindresorhus) [inline test.js:14]:
```suggestion
		if (key !== 'default') {
			continue;
		}
```

to reduce nesting.

**Comment by sindresorhus** [comment]
Oh wow. Haha. I'm sorry you had to suffer through that...

**Comment by sindresorhus** [comment]
> I can also change them to \u+3000 to make it clearer these aren't normal spaces, if you want me to.

👍  Yes

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by nnmrts** [comment]
> Can you fix the merge conflict?

I hope I did this right. I don't know when the last time was I resolved merge conflicts through GitHub.

But the tests seem to pass, everything seems fine, I think this is it. Yay!

**Comment by sindresorhus** [comment]
Woot! 👍🏻

**Comment by SirWindfield** [comment]
The math spinner was really nice :( Maybe I'll go through the suffering to get it back into the package.
